File: 13. Binary Tree/1. Traversals/7. Postorder traversal in a BT/main.cpp
// https://takeuforward.org/data-structure/post-order-traversal-of-binary-tree/


// https://leetcode.com/problems/binary-tree-postorder-traversal/description/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    void inorder(TreeNode* root, vector<int> &arr){
        // If the current node is NULL
        // (base case for recursion), return
        if(root == nullptr){
            return;
        }
        // Recursively traverse the left subtree
        inorder(root->left, arr);
        // Recursively traverse 
        // the right subtree
        inorder(root->right, arr);
        // Push the current node's
        // value into the vector
        arr.push_back(root->val);
    }
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> ans;
        inorder(root, ans);
        return ans;   
    }
};

File: 13. Binary Tree/1. Traversals/7. Postorder traversal in a BT/main.java
import java.util.*;

class Node {
    int data;
    Node left;
    Node right;

    Node(int val) {
        data = val;
        left = null;
        right = null;
    }
}

class Solution {
    public void postorder(Node root, List<Integer> arr) {
        if(root == null) {
            return;
        }

        postorder(root.left, arr);
        postorder(root.right, arr);
        arr.add(root.data);
    }

    public List<Integer> postOrder(Node root) {
        List<Integer> arr = new ArrayList<>();
        postorder(root, arr);

        return arr;
    }
}

public class main {
    public static void main(string[] args) {
        Node root = new Node(1);
        root.left = new Node(2);
        root.right = new Node(3);
        root.left.left = new Node(4);
        root.left.right = new Node(5);
    
        Solution sol = new Solution();
        
        List<Integer> result = sol.postOrder(root);

        System.out.println("Inorder Traversal: ");

        for(int val : result) {
            System.out.print(val + " ");
        }

        System.out.println();
    }
}


/*

time complexity => O(n) => we process each node once fo the traversal

space complexity => O(n) => extra space used foro storing the preorder traversal and the recusrive stack 

*/

